pkg/compact/downsample/streamed_block_writer.go

```diff
@@ -0,0 +1,264 @@
+		}
+		values.add(label.Value)
+	}
+}
+
+// streamedBlockWriter writes downsampled blocks to a new data block. Implemented to save memory consumption
+// by writing chunks data right into the files, omitting keeping them in-memory. Index and meta data should be
+// sealed afterwards, when there aren't more series to process.
+type streamedBlockWriter struct {
+	blockDir       string
+	finalized      bool // set to true, if Close was called
+	logger         log.Logger
+	ignoreFinalize bool // if true Close does not finalize block due to internal error.
+	meta           metadata.Meta
+	totalChunks    uint64
+	totalSamples   uint64
+
+	chunkWriter tsdb.ChunkWriter
+	indexWriter tsdb.IndexWriter
+	indexReader tsdb.IndexReader
+	closers     []io.Closer
+
+	labelsValues labelsValues       // labelsValues list of used label sets: name -> []values.
+	memPostings  *index.MemPostings // memPostings contains references from label name:value -> postings.
+	postings     uint64             // postings is a current posting position.
+}
+
+// NewStreamedBlockWriter returns streamedBlockWriter instance, it's not concurrency safe.
+// Caller is responsible to Close all io.Closers by calling the Close when downsampling is done.
+// In case if error happens outside of the StreamedBlockWriter during the processing,
+// index and meta files will be written anyway, so the caller is always responsible for removing block directory with
+// a garbage on error.
+// This approach simplifies StreamedBlockWriter interface, which is a best trade-off taking into account the error is an
+// exception, not a general case.
+func NewStreamedBlockWriter(
+	blockDir string,
+	indexReader tsdb.IndexReader,
+	logger log.Logger,
+	originMeta metadata.Meta,
+) (w *streamedBlockWriter, err error) {
+	sw := &streamedBlockWriter{
+		logger:       logger,
+		blockDir:     blockDir,
+		indexReader:  indexReader,
+		meta:         originMeta,
+		closers:      make([]io.Closer, 0, 2),
+		labelsValues: make(labelsValues, 1024),
+		memPostings:  index.NewUnorderedMemPostings(),
+	}
+
+	// We should close any opened Closer up to an error.
+	defer func() {
+		if err != nil {
+			var merr tsdb.MultiError
+			merr.Add(err)
+			merr.Add(sw.Close())
+			err = merr.Err()
+		}
+	}()
+
+	sw.chunkWriter, err = chunks.NewWriter(filepath.Join(blockDir, block.ChunksDirname))
+	if err != nil {
+		return nil, errors.Wrap(err, "create chunk streamedBlockWriter")
+	}
+	sw.closers = append(sw.closers, sw.chunkWriter)
+
+	sw.indexWriter, err = index.NewWriter(filepath.Join(blockDir, block.IndexFilename))
+	if err != nil {
+		return nil, errors.Wrap(err, "open index streamedBlockWriter")
+	}
+	sw.closers = append(sw.closers, sw.indexWriter)
+
+	symbols, err := sw.indexReader.Symbols()
+	if err != nil {
+		return nil, errors.Wrap(err, "read symbols")
+	}
+
+	err = sw.indexWriter.AddSymbols(symbols)
+	if err != nil {
+		return nil, errors.Wrap(err, "add symbols")
+	}
+
+	// Stream writer is ready to use.
+	w = sw
```

Comment:
please do not use return arguments in that way. Let's always explicitly return, so `return sw, nil`